convert2rhel/backup.py

```diff
@@ -104,6 +108,150 @@ def restore_pkgs(self):
+    convert2rhel will change on the system which it can restore in case of a failure before the
+    Point-of-no-return (PONR).
+
+    The basic interface to this is a stack.  When a Restorable is pushed onto the stack, it is
+    backed up.  When it is popped off of the stack, it is restored.
+    """
+
+    def __init__(self):
+        self._restorables = []
+
+    def push(self, restorable):
+        """
+        Enable a RestorableChange and track it in case it needs to be restored.
+
+        :arg restorable: RestorableChange object that can be restored later.
+        """
+        if not isinstance(restorable, RestorableChange):
+            raise TypeError("`%s` is not a RestorableChange object" % restorable)
+
+        restorable.enable()
+
+        self._restorables.append(restorable)
+
+    def pop(self):
+        """
+        Restore and then return the last RestorableiChange added to the Controller.
+
+        :returns: RestorableChange object that was last added.
+        :raises IndexError: If there are no RestorableChanges currently known to the Controller.
+        """
+        restorable = self._restorables.pop()
+        restorable.restore()
+
+        return restorable
+
+    def pop_all(self):
+        """
+        Restores all RestorableChanges known to the Controller and then returns them.
+
+        :returns: List of RestorableChange objects that were known to the Controller.
+        :raises IndexError: If there are no RestorableChanges currently known to the Controller.
+
+        After running, the Controller object will not know about any RestorableChanges.
+        """
+        restorables = self._restorables
+
+        if not restorables:
+            raise IndexError("pop from empty list")
+
+        # We want to restore in the reverse order the changes were enabled.
+        for restorable in reversed(restorables):
+            restorable.restore()
+
+        # Reset the internal storage in case we want to use it again
+        self._restorables = []
+
+        # Now that we know everything succeeded, reverse the list that we return to the user
+        restorables.reverse()
+
+        return restorables
+
+
+@six.add_metaclass(abc.ABCMeta)
+class RestorableChange(object):
+    """
+    Interface definition for types which can be restored.
+    """
+
+    @abc.abstractmethod
+    def __init__(self):
+        self.enabled = False
+
+    @abc.abstractmethod
+    def enable(self):
+        """
+        Backup should be idempotent.  In other words, it should know if the resource has already
+        been backed up and refuse to do so a second time.
+        """
+        self.enabled = True
+
+    @abc.abstractmethod
+    def restore(self):
+        """
+        Restore the state of the system.
+        """
+        self.enabled = False
+
+
+class RestorableRpmKey(RestorableChange):
+    def __init__(self, keyfile):
+        super(RestorableRpmKey, self).__init__()
+        self.previous_state = None
```

Comment:
So, I'm not a big fan of having states as a normal string in the code and not having a good way to verify/assert that the state is actually correct… I don't know how you feel about it, but to me, it seems that this can be error-prone so easily. 

The reason is that I'm imagining that someone could change these states at will in any other PR and introduce new ones, or just mess up the name of the state and lose a lot of time debugging or figuring out why. Another possibility would be someone using this class as a subclass of something else and introducing other states, and we would need to keep track of that in two separate classes. _I know, I went too far in the imagination section, but…_

Since we have two states for now (`not_installed` | `installed`), it seems to be that we could replace that with a boolean property just fine.  Maybe, something like that could work: 

```suggestion
        self.is_installed = False
```

or

```suggestion
        self.is_rpm_installed = False
```

I would say that if we want to keep the states as strings here, I would come up with something to assert that the values can't be different from what we propose/expect here… 

Let me know what you think about this… If you don't think that it makes sense, feel free to just resolve the comment!

Reply:
I think that Enums would be preferred in the general case of more than three values but we don't have access to Enums until we can run only on Python3 (python-3.4 specifically).  Until then, for more than three values, I think strings should be preferred.  They do suffer from potential typos as you point out, but the other side of that coin is that, they are easily extensible.  And strings are just better than any of the other choices until Enums were added.

In this case, I think you're right to think we have less values so a boolean might be okay.  However:
* This is actually a tristate, not a True|False.  The third state is "unknown". (represented by None).
* I don't think we need to use some sort of temporal reference in the variable name.  This is not about whether the gpg key is imported now.  it's about whether the gpg  key was imported prior to our performing any work on the system.

So maybe something like `self.previously.installed = (True|False|None)`

Reply:
Switched to `self.previously_installed` and refactored the logic in `restore()` to be clearer.

Reply:
Like the modifications! Thanks for the explanation as well.